src/types.rs

```diff
@@ -268,24 +276,103 @@ impl Type {
+            TypeEnum::Prim(PrimType::Extension(custy)) => {
+                custy.validate(extension_registry, type_vars)
+            }
+            TypeEnum::Prim(PrimType::Function(ft)) => ft.validate(extension_registry, type_vars),
+            TypeEnum::Prim(PrimType::Variable(idx, bound)) => {
+                check_typevar_decl(type_vars, *idx, &TypeParam::Type(*bound))
+            }
+        }
+    }
+
+    /// Substitute the specified [TypeArg]s for type variables in this type.
+    ///
+    /// # Arguments
+    ///
+    /// * `args`: values to substitute in; there must be at least enough for the
+    /// typevars in this type (partial substitution is not supported).
+    ///
+    /// * `extension_registry`: for looking up [TypeDef]s in order to recompute [TypeBound]s
+    /// as these may get narrower after substitution
+    ///
+    /// # Panics
+    ///
+    /// If a [TypeArg] (that is referenced by a typevar in this type) does not contain a [Type],
+    /// contains a type with an incorrect [TypeBound], or there are not enough `args`.
+    /// These conditions can be detected ahead of time by [Type::validate]ing against the [TypeParam]s
+    /// and [check_type_args]ing the [TypeArg]s against the [TypeParam]s.
+    pub(crate) fn substitute(&self, exts: &ExtensionRegistry, args: &[TypeArg]) -> Self {
+        match &self.0 {
+            TypeEnum::Prim(PrimType::Alias(_)) | TypeEnum::Sum(SumType::Simple { .. }) => {
+                self.clone()
+            }
+            TypeEnum::Prim(PrimType::Variable(idx, bound)) => match args.get(*idx) {
+                Some(TypeArg::Type { ty }) => ty.clone(),
+                Some(v) => panic!(
+                    "Value of variable {:?} did not match cached param {}",
+                    v, bound
+                ),
+                None => panic!("No value found for variable"), // No need to support partial substitution for just type schemes
+            },
+            TypeEnum::Prim(PrimType::Extension(cty)) => {
+                Type::new_extension(cty.substitute(exts, args))
+            }
+            TypeEnum::Prim(PrimType::Function(bf)) => Type::new_function(bf.substitute(exts, args)),
+            TypeEnum::Tuple(elems) => Type::new_tuple(subst_row(elems, exts, args)),
+            TypeEnum::Sum(SumType::General { row }) => Type::new_sum(subst_row(row, exts, args)),
+        }
+    }
+}
+
+fn subst_row(row: &TypeRow, exts: &ExtensionRegistry, args: &[TypeArg]) -> TypeRow {
+    let res = row
+        .iter()
+        .map(|t| t.substitute(exts, args))
+        .collect::<Vec<_>>()
+        .into();
+    res
+}
+
+pub(crate) fn check_typevar_decl(
+    decls: &[TypeParam],
+    idx: usize,
+    used_as: &TypeParam,
+) -> Result<(), SignatureError> {
+    match decls.get(idx) {
+        None => Err(SignatureError::FreeTypeVar {
+            idx,
+            num_decls: decls.len(),
+        }),
+        Some(decl) => {
+            if decl == used_as {
```

Comment:
Yeah this is a bit tricky. The field inside `TypeArgVariable` is intended as a strict cache of the declaration, and then we check it fits (with contains) in `check_type_arg`. So, the `cached_decl` is the widest thing that fits within the decl. Alternatively, we could require equality in `check_type_arg`, and have a containment check here; again, that would mean there was exactly one correct value for the cached_decl, just the opposite way around (it would be the narrowest thing that contains the usage, essentially).

Or we could even allow containment at both ends, and then `cached_decl` could be *anything* that fits in the decl and contains the usage - there could be multiple "legal" values at each usage site. I dislike the latter just because it's less precise...

The *ideal* would be to avoid `cached_decl` altogether - which would be possible, but every API that takes `TypeArg`s would have to have an extra parameter (the `TypeParam` decl's of any variable that might occur in those `TypeArg`s). Specifically, that includes `check_type_arg`(`s`) and friends, which already take a `TypeParam` (the thing to check whether the TypeArg fits into), so then check_type_arg(s) would take two TypeParam-things, which invites confusion as to which is which....and `check_type_arg`(`s`) is used in quite a lot of places, so sadly I think I have to dismiss that one :-(.

Reply:
I've done a bit of renaming, hopefully that's clearer. I can add a comment here (that the containment check is done in `check_type_arg`) if you think that'd help?

Reply:
I see, the renaming makes things more clear 👍 I think adding a comment pointing to `check_type_arg` wouldn't be a bad idea

Reply:
Ok, done